added custom watermark for kinesis reader

sjmittal opened the pull request:
The PR is for attempted fix of the issue #28760 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

damccorm commented:
@sjmittal what are next steps here? It looks like currently this PR is not compiling

damccorm commented:
@damondouglas @Abacn @ahmedabu98 @johnjcasey could someone please take a look at this one?

johnjcasey commented:
Ack, taking a look

johnjcasey commented on sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/EFOShardSubscribersPool.java:
```diff
@@ -331,6 +331,15 @@ Instant getWatermark() {
+    if (sharedStates.isEmpty()) {
+      return null;
+    }
+    return sharedStates.get(0).getEventTime(record);
```
Is it valid to get the event time from any shard state?

sjmittal replied:
We get watermark also from `ShardState#getWatermark` so it seems logical that event timestamp is also fetched from `ShardState#getEventTime`.

johnjcasey replied:
SGTM

johnjcasey commented on sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/ShardReadersPool.java:
```diff
@@ -255,6 +256,16 @@ Instant getLatestRecordTimestamp() {
+    if (shardRecordsIterators.isEmpty()) {
+      return null;
+    }
+    return shardRecordsIterators.get(0).getEventTime(record);
```
Same question here, is it valid to ask for the event time from any shard record iterator?

sjmittal replied:
We get watermark also from `ShardRecordsIterator#getWatermark` so it seems logical that event timestamp is also fetched from `ShardRecordsIterator#getEventTime` which internally returns `watermarkPolicy.getTimestamp`.

johnjcasey replied:
SGTM

johnjcasey reviewed (changes requested):
We need to add tests to this change

sjmittal commented:
> We need to add tests to this change

I have updated the test cases to `org.apache.beam.sdk.io.aws2.kinesis.WatermarkPolicyTest`.
The essence of this PR is to get the current timestamp of the stream from the record if custom watermark policy is set for it, otherwise if falls back to the arrival time policy.
Hence I have updated the test cases for above class. 

In other test classes where we test the Kinesis reader or shard subscribers the setup itself of these test cases assumes arrival time policy hence there are no changes to these test classes.

damccorm commented:
@johnjcasey could you please take another look at this one/help take this pr to completion?

damccorm commented:
@johnjcasey could you please take another look here?
